### PR TITLE
functionalize(): make "additionally removing views" toggleable

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR goes with the core-side change here: https://github.com/pytorch/pytorch/pull/75302, which updates the functionalization pass to be able to turn view ops into view_copy ops. Mobile mentioned that they would like to be able to use the `functionalize()` transform to trace models for running on mobile, and removing view ops (cc @ZolotukhinM)

This PR updates `functionalize()` to be toggleable: `functionalize(remove='mutations')` (the default) will remove mutations but preserve views. `functionalize(remove='mutations_and_views')` will remove mutations, and additionally convert `view` operators into their corresponding `view_copy` operators.

Some extra stuff in the PR:
- added more testing for out= mutations, which I fixed in a corresponding core PR: https://github.com/pytorch/pytorch/pull/75818
- remove an unnecessary "regenerate_from_base()" call that gives you an extra view_copy call in some cases. Goes with core PR: https://github.com/pytorch/pytorch/pull/75819

### CODE REVIEW

File: functorch/csrc/DynamicLayer.cpp
Comment (bdhirsh) [inline functorch/csrc/DynamicLayer.cpp:562]: Apparently you need to compile with the `-Werror=return-type` flag, or else you get silent UB if you forget to return from a non-void function (aka this bug) :(

File: functorch/csrc/init.cpp
Comment (bdhirsh) [inline functorch/csrc/init.cpp:147]: sorry for missing this the first time around :)

File: functorch/csrc/DynamicLayer.cpp
Comment (zou3519) [reply functorch/csrc/DynamicLayer.cpp:574]: :(, we used to have -Werror turned on for all errors but it turns out PyTorch doesn't and PyTorch would introduce warnings

File: functorch/_src/eager_transforms.py
Comment (zou3519) [inline functorch/_src/eager_transforms.py:1219]: From a user's perspective: they don't know what it means to "reapply views" because they don't have the context that  functionalization involves (1) removing views and mutations and then (2) reapplying the views.

Would `functionalize(func, *, remove_views)` be a more descriptive name?

Comment (zou3519) [reply functorch/_src/eager_transforms.py:1219]: Regarding the default: I think of functionalization as "remove in-place operations only", so "removing views" seems like an additional feature on top. Because of that, I think the default should be "remove in-place operations and reapply views".

File: functorch/csrc/init.cpp
Comment (zou3519) [inline functorch/csrc/init.cpp:160]: nit: we can save ourselves a bit of repetition:
```
optional<IncludeDispatchKeyGuard> guard;
if (add_back_views) {
  guard.emplace(c10::DispatchKey::FunctionalizeAddBackViews);
}
auto any_updates = functional->apply_updates();
if (any_updates) {
  functional->regenerate_from_base();
}
```

File: functorch/csrc/DynamicLayer.h
Comment (zou3519) [inline functorch/csrc/DynamicLayer.h:51]: Lol I am going to turn these optional fields into a Union after we merge this, I didn't expect us to add so many

File: test/test_eager_transforms.py
Comment (zou3519) [inline test/test_eager_transforms.py:2779]: file an issue?

File: test/test_eager_transforms.py
Comment (zou3519) [inline test/test_eager_transforms.py:2863]: You sure we don't need to test CUDA?

**Review by zou3519 — COMMENTED** [review]
Code LGTM.

This needs a rebase and I want to bikeshed the API and defaults a little more (maybe there shouldn't be a default yet)

File: functorch/_src/eager_transforms.py
Comment (bdhirsh) [reply functorch/_src/eager_transforms.py:1219]: That all sounds totally reasonable to me:

(1) default is "only removing inplace"
(2) string argument instead of bool, with a nice descriptive error message

File: test/test_eager_transforms.py
Comment (bdhirsh) [reply test/test_eager_transforms.py:2863]: There really.... shouldn't be anything cuda-specific about anything going on. I took it out because my local `pytest` is broken and I was too lazy to fix it, and kept forgetting the right "_CPU" incantation to add to the end of my tests when running locally lol. But the tests don't take too long to run, so I may as well add it back.

File: test/test_eager_transforms.py
Comment (bdhirsh) [reply test/test_eager_transforms.py:2784]: done: https://github.com/pytorch/pytorch/issues/75909

**Comment by bdhirsh** [comment]
Pushed some more changes. Updates:

(1) API bikeshed. I liked your suggestion of `functionalize(remove='mutations|mutations_and_views')`, where you get a nice error message otherwise (and we default to the functorch default which is mutations)

(2) I ended up killing the `FunctionalizeAddBackViews` dispatch key, and used an extra piece of TLS instead. This saves us a dispatch key.

**Review by zou3519 — APPROVED** [review]
LGTM, you probably want to change the PR body (it still mentions FunctionalizeAddBackViews). Also it would be good to wait until the CI turns green

**Comment by zou3519** [comment]
@bdhirsh btw, functorch CI runs off of the PyTorch nightly binary, so we have two options:
1. Wait until your pytorch-side change makes it to PyTorch nightlies (hopefully this will happen tonight)
2. Yolo merge this and break functorch CI

**Comment by bdhirsh** [comment]
Lol you probably saw me blindly kicking off CI again hoping something would happen (although at least `python test/test_eager_transforms.py` passes for me locally).

I'm ok with either, depending on how urgently @ZolotukhinM would like this change for mobile

**Comment by bdhirsh** [comment]
CI is red, but I think I'm seeing the same set of test failures on main: https://app.circleci.com/pipelines/github/pytorch/functorch/2432/workflows/4657bb2a-cba6-4ce4-b409-e72e5229e3c4/jobs/14516/tests

**Comment by zou3519** [comment]
@bdhirsh feel free to merge, the errors look pre-existing and we'll figure them out as we go along today

**Pull request merged**
